# GPS Unified Pipeline - Work-in-progress

GPS Unified Pipeline is a Nextflow Pipeline for processing Streptococcus pneumoniae sequencing raw reads (FASTQ files) by the GPS Project ([Global Pneumococcal Sequencing Project](https://www.pneumogen.net/gps/)). 

&nbsp;
## Current workflow and progress
![Workflow](doc/workflow.drawio.svg)

&nbsp;
## Usage
### Requirement
- Conda / Mamba
- Git
### Setup
1. Clone the repository
    ```
    git clone https://github.com/HarryHung/gps-unified-pipeline.git
    ```
2. Go into the local copy of the repository
    ```
    cd gps-unified-pipeline
    ```
3. Setup Conda Environment (If using Mamba, replace `conda` with `mamba` in the following commands)
   - MacOS (Intel CPU)
     ```
     conda env create -f environment_mac.yml
     conda activate pipeline
     ```
   - MacOS (Apple Silicon)
     ```
     CONDA_SUBDIR=osx-64 conda env create -f environment_mac.yml
     conda activate pipeline
     conda config --env --set subdir osx-64
     ```
   - Linux
     ```
     conda env create -f environment_linux.yml
     conda activate pipeline
     ```

### Run
- You can run the pipeline without arguments/parameters. It will attempt to get the raw reads from the default location (`input` directory inside the `gps-unified-pipeline` local repository)
  ```
  nextflow run main.nf
  ```
- You can also specific the location of the raw reads by adding argument `--reads`
  ```
  nextflow run main.nf --reads /path/to/raw-reads-directory
  ```
- For a test run, you could use the included test reads in the `test_input` directory
  ```
  nextflow run main.nf --reads test_input
  ```
  - `9870_5#52` will fail the Taxonomy QC and hence Overall QC, therefore without analysis results
  - `21127_1#156` should pass Overall QC, and with analysis results

### Output
- By default, the pipeline outputs the results into `output` directory inside the `gps-unified-pipeline` local repository
- It can be changed by adding argument `--output`
  ```
  nextflow run main.nf --output /path/to/output-directory
  ```
- The following directories and files are output into the output directory
  | Directory / File | Description |
  | --- | ---|
  | assemblies | This directory contains all assemblies (`.fasta`) generated by the pipeline |
  | summary.csv | This file contains all the information generated by the pipeline on each sample |